initialize current linearization point when we refine the mesh

jdannberg opened the pull request:
Possible fix for #5042. See discussion there. 

* [x] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).

### For new features/models or changes of existing features:

* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](http://www.math.clemson.edu/~heister/manual.pdf#sec%3Awriting_tests) for the new feature/benchmark in the [tests/](../blob/main/tests/) directory.
* [x] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

If this seems reasonable, I can add the test from #5042.

gassmoeller commented:
I think this is almost the correct solution. I had to dig a bit to find out why this is causing a problem, but here is what I gathered:
- When `setup_dofs()` is called after mesh refinement all solution vectors and current_initialization_point are reinitialized to the new dof structure and set to 0.0 (has to be done, because the vector entries no longer correspond to the current grid)
- `solution` and `old_solution` are immediately set to the new interpolated values, but `current_linearization_point` is only set to the new values the next time `solve_timestep` is called (which calls `initialize_current_linearization_point`), which is at the start of the next timestep.
- This is a problem, because `replace_outflow_boundary_ids()` which is part of `compute_current_constraints()` evaluates the current linearization point to determine outflow boundaries before `solve_timestep` reinitializes the vector to the correct values, therefore it always sees a velocity of 0 at the boundary after mesh refinement.

I think a better solution than the current one would be to do this:
```
current_linearization_point = distributed_system;
```
which would set `current_linearization_point` to exactly the value of `solution` (instead of an extrapolation from `old_solution` and `old_old_solution`). This is the natural state of `current_linearization_point` after a timestep anyway (i.e. the state it would have in a timestep without mesh refinement). It will then be reset to a new extrapolation for the next timestep at the usual time after the boundary ids have been reset.

Can you give this a try? If it works, feel free to add a test.

jdannberg commented:
I have changed the code as you suggested (which works), and as discussed, also modified the `replace_outflow_boundary_ids()` function so that zero flow counts as outflow, which fixes the problem with time step 0. 

I have also added a test and changelog entry.

jdannberg commented:
I have updated the test results. 
I have looked through the list of tests, and all of the ones that change have a fixed boundary composition at a model boundary where the velocity is prescribed. In some of these, the prescribed velocity is zero (so that the boundary composition was prescribed before over the whole model run and now isn't any more), in others, the prescribed composition is different than the initial composition (so that the composition is different in time step zero because we now no longer apply the composition boundary conditions before the velocity solution is calculated). 
So I think it makes sense that all these tests change.

tjhei commented on source/simulator/core.cc:
```diff
@@ -1746,6 +1746,8 @@ namespace aspect
       constraints.distribute (old_distributed_system);
       old_solution = old_distributed_system;
 
+      current_linearization_point = distributed_system;
```
Can you leave a comment here that summarizes the discussion (that we need some value for the inflow/outflow determination and that it will be updated later)

tjhei commented on source/simulator/helper_functions.cc:
```diff
@@ -2115,7 +2115,8 @@ namespace aspect
                 // ... and change the boundary id of any outflow boundary faces.
-                if (integrated_flow > 0)
+                // If there is no flow, we do not want to apply dirichlet boundary conditions either.
+                if (integrated_flow >= 0)
```
Yes, note that this is now the same as in DG:
https://github.com/geodynamics/aspect/blob/cbc7833440cd05c7370bc28f4119199f052dbcfa/source/simulator/assemblers/advection.cc#L780

jdannberg replied:
Oh, it's great that it's now consistent between the two methods.

tjhei reviewed (approved):
Very nice work!

jdannberg commented:
Thank you! I addressed your one comment.

jdannberg commented:
...and I have no idea why a test is failing, since all I changed was a comment.

gassmoeller commented:
/rebuild
